### PR TITLE
DEPR: GH10623 remove items from msgpack.encode for blocks

### LINKED ISSUE #10623: DEPR: msgpack encode 'items'
At some point deprecate `items` in `msgpack.encode()` for blocks. xref #10527

### PR DESCRIPTION
**Pull request by kawochen**
closes #10623

### CODE REVIEW

**Comment by kawochen** [comment]
hmm legacy file for 0.17.1 not passing.  I think it's tz-related.  Will check later.

**Comment by kawochen** [comment]
passing now.  had to make changes for `DatetimeTZBlock` (it wasn't tested?) and skip files packed in P2 with 0.17 :( due to #12142

File: pandas/io/packers.py [hunk 1]
```diff
@@ -371,9 +385,8 @@ def encode(obj):
             return {'typ': 'block_manager',
                     'klass': obj.__class__.__name__,
                     'axes': data.axes,
-                    'blocks': [{'items': data.items.take(b.mgr_locs),
-                                'locs': b.mgr_locs.as_array,
-                                'values': convert(b.values),
+                    'blocks': [{'locs': b.mgr_locs.as_array,
+                                'values': convert_if_not_dt_index(b.values),
```
Comment (jreback) [inline pandas/io/packers.py]: this is way hacky. you shouldn't need to specifically do this (test for ``DatetimeIndex``. It *should* just be converted normally. The entire difference here is the dispatch on the dtype, which in this case would be a string.

Comment (kawochen) [inline pandas/io/packers.py]: if it gets converted to a numpy array now, wouldn't the whole `encode`/`decode` for `DatetimeIndex` be skipped?  Or is there a way to construct DatetimeTZBlock from numpy arrays?

Comment (jreback) [inline pandas/io/packers.py]: actually I think you could trivially change ``core/internals.py/make_block`` to take a ``DatetimeTZDtype`` and correctly create it. Then you would have to make to serialize that properly (as a string is prob ok) I think (you would need to fix the reverse lookup as it only deals with numpy atm). 

I think the same problem exists with categorical, prob not testing it.

**Comment by kawochen** [comment]
updated

File: pandas/core/internals.py [hunk 2]
```diff
@@ -2094,9 +2094,17 @@ class DatetimeTZBlock(NonConsolidatableMixIn, DatetimeBlock):
...
     def __init__(self, values, placement, ndim=2, **kwargs):
+        # ndarray values must have dtype
+        if isinstance(values, np.ndarray):
+            values = self._holder(values)
+            dtype = kwargs.pop('dtype')
+            if isinstance(dtype, compat.string_types):
+                dtype = DatetimeTZDtype.construct_from_string(dtype)
+            values = values.tz_localize('UTC').tz_convert(dtype.tz)
 
         if not isinstance(values, self._holder):
```
Comment (jreback) [inline pandas/core/internals.py]: you should do this first, then you can coerce if you have a dtype

File: pandas/core/internals.py [hunk 3]
```diff
@@ -2426,7 +2434,9 @@ def make_block(values, placement, klass=None, ndim=None, dtype=None,
             klass = CategoricalBlock
         else:
             klass = ObjectBlock
-
+    if klass is DatetimeTZBlock and not is_datetimetz(values):
```
Comment (jreback) [inline pandas/core/internals.py]: hmm, this should be in the above paths as well, no?

**Comment by jreback** [comment]
fyi some of these fixes will prob address: https://github.com/pydata/pandas/issues/11455 (and linked issues)

**Comment by kawochen** [comment]
updated... but I forgot to squash :(  will do that later

File: pandas/io/packers.py [hunk 4]
```diff
@@ -245,6 +246,10 @@ def unconvert(values, dtype, compress=None):
     if dtype == np.object_:
         return np.array(values, dtype=object)
 
+    if isinstance(dtype, string_types):
```
Comment (jreback) [inline pandas/io/packers.py]: I get that you have to do this, we prob need something like (in core/common.py)


```
def pandas_dtype(dtype):
     if isinstance(dtype, compat.string_types):
            try:
                return DatetimeTZDtype.construct_from_string(dtype)
            except TypeError:
                pass

            try:
               return CategoricalDtype.construct_from_string(dtype)
            except TypeError:
               pass

           dtype = np.dtype(dtype)

    return dtype
```

**Comment by jreback** [comment]
pls add a whatsnew as well (put in API changes).
make a note in ``io.rst`` that the writing format is changing in 0.18.0 so we won't have forward compat (IOW older version of pandas cannot read newer version)

**Comment by jreback** [comment]
this would close #12142 as well? (should add a warning in the docs about this) (e.g. show your matrix)

**Comment by kawochen** [comment]
I can change all the keys to unicode, and then it will.  will update tonight

**Comment by kawochen** [comment]
Ah half way through but have to go.  I will get it done tonight but is that too late? :(

**Comment by jreback** [comment]
sure no prob

**Comment by kawochen** [comment]
updated.  manually tested with python 2.7 and 3.5 on linux (zlib and no compress).

File: pandas/io/packers.py [hunk 5]
```diff
@@ -265,28 +269,31 @@ def unconvert(values, dtype, compress=None):
     return np.fromstring(values, dtype=dtype)
 
 
+def _u(x):
```
Comment (jreback) [inline pandas/io/packers.py]: just use ``u()`` from ``compat/__init__.py`` (and amend with this).

File: doc/source/whatsnew/v0.18.0.txt [hunk 6]
```diff
@@ -507,6 +507,32 @@ Subtraction by ``Timedelta`` in a ``Series`` by a ``Timestamp`` works (:issue:`1
...
+      | Packed with          | Can be unpacked with   |
+      +======================+========================+
+      | pre-0.17 / Python 2  | any                    |
+      +----------------------+------------------------+
+      | pre-0.17 / Python 3  | any                    |
+      +----------------------+------------------------+
+      | 0.17 / Python 2      | - 0.17 / Python 2      |
+      |                      | - >=0.18 / any Python  |
+      +----------------------+------------------------+
+      | 0.17 / Python 3      | >=0.18 / any Python    |
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: so maybe say (and in io.rst as well). that 0.18.0 is backward compat for reading (except for 0.17/py2)

File: pandas/io/packers.py [hunk 7]
```diff
@@ -554,25 +561,25 @@ def create_block(b):
...
             try:
-                return dtype(obj['data'])
+                return dtype(obj[u'data'])
             except:
-                return dtype.type(obj['data'])
-    elif typ == 'np_complex':
-        return complex(obj['real'] + '+' + obj['imag'] + 'j')
+                return dtype.type(obj[u'data'])
+    elif typ == u'np_complex':
+        return complex(obj[u'real'] + u'+' + obj[u'imag'] + u'j')
```
Comment (jreback) [inline pandas/io/packers.py]: i'll put some additional test files for macosx/windows. you want 0.17.1 & 0.17.0?

**Comment by kawochen** [comment]
updated.  So this is how I tested P2 & P3 compatibility: use the script to generate legacy data for 0.18.0 in P2 and P3, put them in 0.18.0 folder, and let the tests pick them up.

**Comment by jreback** [comment]
thanks!

I added some more test msgpacks for 0.17.1 to cover the bases

**Comment by jreback** [comment]
https://travis-ci.org/pydata/pandas/jobs/109620001

I think we need to skip if blosc is not installed on these tests. 

can you do a PR for this?

**Pull request closed without merging**